fluss-client/src/main/java/com/alibaba/fluss/client/table/writer/UpsertWriterImpl.java

```diff
@@ -126,15 +143,19 @@ public CompletableFuture<UpsertResult> upsert(InternalRow row) {
      */
     public CompletableFuture<DeleteResult> delete(InternalRow row) {
         byte[] key = primaryKeyEncoder.encode(row);
-        byte[] bucketKey =
-                bucketKeyEncoder == primaryKeyEncoder ? key : bucketKeyEncoder.encode(row);
+        int bucketId = getBucketId(key, row);
         return send(new WriteRecord(
-                        getPhysicalPath(row),
-                        WriteKind.DELETE,
-                        key,
-                        bucketKey,
-                        null,
-                        targetColumns))
+                        getPhysicalPath(row), WriteKind.DELETE, key, bucketId, null, targetColumns))
                 .thenApply(ignored -> DELETE_SUCCESS);
     }
+
+    private int getBucketId(byte[] keyBytes, InternalRow row) {
+        if (isDefaultBucketKey && bucketAssigner instanceof HashBucketAssigner) {
+            // is is default bucket key, and assigner is HashBucketAssAssigner,
+            // we can reuse the primary key bytes to calculate bucket id
+            return ((HashBucketAssigner) bucketAssigner).assignBucket(keyBytes);
```

Comment:
This is a super hack to do a hard cast, it means we didn't have a good abstraction here. 

As suggested above, if all the encoded (primary/bucket) key use the datalake KeyEncoder, I think the parameter of `BucketAssigner` can be changed to `byte[] key` from the `InternalRow`.